### ISSUE TITLE
Test failure(s): automation.module.timer.internal.RuntimeRuleTest

### ISSUE THREAD

**Issue by maggu2810** [eclipse/smarthome#2057]
Related to: #1757, #2054

The last changes shows the error in more details.
The rule itself is using wildcards for the cron job timer.
The rule should be disabled but is not disabled after 10 seconds.
I don't know if the rule is triggered such frequently that the disabling does not work anymore.
@danchom Could you give me some informations why the disable does not work?

```text
-------------------------------------------------------
 T E S T S
-------------------------------------------------------
Running org.eclipse.smarthome.automation.module.timer.internal.RuntimeRuleTest
16:51:26.895 [main] DEBUG o.e.s.c.c.registry.AbstractRegistry - Provider 'com.sun.proxy.$Proxy7' has been added.
16:51:26.900 [main] INFO  o.e.s.a.c.internal.RuleRegistryImpl - Rule Managed Provider: org.eclipse.smarthome.automation.core.internal.ManagedRuleProvider@aafcffa is added.
16:51:26.900 [main] INFO  o.e.s.a.c.internal.RuleRegistryImpl - Rule provider: org.eclipse.smarthome.automation.core.internal.ManagedRuleProvider@aafcffa is added.
16:51:26.903 [main] DEBUG o.e.s.c.c.registry.AbstractRegistry - Provider 'org.eclipse.smarthome.automation.core.internal.ManagedRuleProvider' has been added.
16:51:26.933 [main] DEBUG o.e.s.a.i.c.p.AutomationResourceBundlesEventQueue - Process bundle event 2, for automation bundle 'org.eclipse.smarthome.automation.module.core' 
16:51:26.935 [main] DEBUG o.e.s.a.i.c.p.AutomationResourceBundlesEventQueue - Process bundle event 2, for automation bundle 'org.eclipse.smarthome.automation.module.timer' 
16:51:26.960 [Timer-0] DEBUG org.quartz.utils.UpdateChecker - Checking for available updated version of Quartz...
16:51:26.964 [Automation Provider Processing Queue] DEBUG o.e.s.a.i.c.p.RuleResourceBundleImporter - Parse rules from bundle 'org.eclipse.smarthome.automation.module.core' 
16:51:26.965 [main] DEBUG o.e.s.c.c.registry.AbstractRegistry - Provider 'org.eclipse.smarthome.core.items.ManagedItemProvider' has been added.
16:51:26.966 [Automation Provider Processing Queue] DEBUG o.e.s.a.i.c.p.RuleResourceBundleImporter - Parse rules from bundle 'org.eclipse.smarthome.automation.module.timer' 
16:51:27.045 [main] INFO  o.e.s.a.m.t.internal.RuntimeRuleTest - Check auto update
16:51:27.059 [EventAdmin Async Event Dispatcher Thread] DEBUG o.e.s.core.common.ThreadPoolManager - Created thread pool 'safeCall' with size 5
16:51:27.061 [safeCall-2] DEBUG o.e.s.core.common.ThreadPoolManager - Created thread pool 'items' with size 5
16:51:27.226 [main] INFO  o.e.s.a.m.t.internal.RuntimeRuleTest - Auto update works
16:51:27.226 [main] INFO  o.e.s.a.m.t.internal.RuntimeRuleTest - Create rule
16:51:27.249 [main] INFO  o.e.s.a.m.t.internal.RuntimeRuleTest - Rule created: MyRule-1729261805
16:51:27.250 [main] INFO  o.e.s.a.m.t.internal.RuntimeRuleTest - Add rule
16:51:27.257 [main] DEBUG o.e.s.a.core.internal.RuleEngine - Added rule 'MyRule-1729261805'
16:51:27.274 [main] INFO  org.quartz.core.QuartzScheduler - Scheduler DefaultQuartzScheduler_$_NON_CLUSTERED started.
16:51:27.274 [DefaultQuartzScheduler_QuartzSchedulerThread] DEBUG o.quartz.core.QuartzSchedulerThread - batch acquisition of 0 triggers
16:51:27.279 [DefaultQuartzScheduler_QuartzSchedulerThread] DEBUG o.quartz.core.QuartzSchedulerThread - batch acquisition of 1 triggers
16:51:27.279 [main] DEBUG o.e.s.c.c.r.AbstractManagedProvider - Added new element MyRule-1729261805 to ManagedRuleProvider.
16:51:27.279 [main] INFO  o.e.s.a.m.t.internal.RuntimeRuleTest - Rule added
16:51:27.280 [main] INFO  o.e.s.a.m.t.internal.RuntimeRuleTest - Enable rule and wait for idle status
16:51:27.281 [main] INFO  o.e.s.a.core.internal.RuleEngine - The rule rId = MyRule-1729261805 is already enabled
16:51:27.283 [DefaultQuartzScheduler_QuartzSchedulerThread] DEBUG o.q.simpl.PropertySettingJobFactory - Producing instance of Job 'DEFAULT.TimerTrigger267908ab-a8cc-41b0-a492-9ad82d2593d7', class=org.eclipse.smarthome.automation.module.timer.handler.CallbackJob
NONE
16:51:27.291 [DefaultQuartzScheduler_Worker-1] DEBUG org.quartz.core.JobRunShell - Calling execute on job DEFAULT.TimerTrigger267908ab-a8cc-41b0-a492-9ad82d2593d7
16:51:27.292 [DefaultQuartzScheduler_Worker-1] DEBUG o.e.s.a.core.internal.RuleEngine - The trigger 'MyTimerTrigger' of rule 'MyRule-1729261805' is triggred.
16:51:27.292 [DefaultQuartzScheduler_QuartzSchedulerThread] DEBUG o.quartz.core.QuartzSchedulerThread - batch acquisition of 1 triggers
16:51:27.295 [main] INFO  o.e.s.a.m.t.internal.RuntimeRuleTest - Rule is enabled and idle
16:51:27.296 [main] INFO  o.e.s.a.m.t.internal.RuntimeRuleTest - Disable rule
16:51:27.296 [pool-2-thread-1] DEBUG o.e.s.a.m.c.h.ItemStateConditionHandler - ItemStateCondition 'MyItemStateCondition'checking if myLampItem (State=OFF) = OFF
16:51:27.296 [pool-2-thread-1] DEBUG o.e.s.a.m.c.h.ItemStateConditionHandler - ConditionSatisfied --> true
16:51:27.297 [pool-2-thread-1] DEBUG o.e.s.a.m.c.h.ItemPostCommandActionHandler - Executing ItemPostCommandAction on Item myLampItem with Command ON
16:51:27.298 [DefaultQuartzScheduler_QuartzSchedulerThread] DEBUG o.quartz.core.QuartzSchedulerThread - batch acquisition of 0 triggers
16:51:27.299 [pool-2-thread-1] DEBUG o.e.s.a.core.internal.RuleEngine - The rule 'MyRule-1729261805' is executed.
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
NONE
16:51:37.421 [main] DEBUG o.e.s.c.c.registry.AbstractRegistry - Provider '$Proxy7' has been removed.
16:51:37.422 [main] INFO  o.e.s.a.c.internal.RuleRegistryImpl - Rule Managed provider: org.eclipse.smarthome.automation.core.internal.ManagedRuleProvider@aafcffa is removed.
16:51:37.424 [main] DEBUG o.e.s.c.c.registry.AbstractRegistry - Provider 'ManagedItemProvider' has been removed.
16:51:37.432 [main] DEBUG o.e.s.c.c.registry.AbstractRegistry - Provider 'com.sun.proxy.$Proxy7' has been added.
16:51:37.433 [main] INFO  o.e.s.a.c.internal.RuleRegistryImpl - Rule Managed Provider: org.eclipse.smarthome.automation.core.internal.ManagedRuleProvider@76ba13c is added.
16:51:37.433 [main] INFO  o.e.s.a.c.internal.RuleRegistryImpl - Rule provider: org.eclipse.smarthome.automation.core.internal.ManagedRuleProvider@76ba13c is added.
16:51:37.433 [main] DEBUG o.e.s.c.c.registry.AbstractRegistry - Provider 'org.eclipse.smarthome.automation.core.internal.ManagedRuleProvider' has been added.
16:51:37.435 [main] DEBUG o.e.s.c.c.registry.AbstractRegistry - Provider 'org.eclipse.smarthome.core.items.ManagedItemProvider' has been added.
16:51:37.448 [main] DEBUG o.e.s.c.c.registry.AbstractRegistry - Provider '$Proxy7' has been removed.
16:51:37.449 [main] INFO  o.e.s.a.c.internal.RuleRegistryImpl - Rule Managed provider: org.eclipse.smarthome.automation.core.internal.ManagedRuleProvider@76ba13c is removed.
16:51:37.450 [main] DEBUG o.e.s.c.c.registry.AbstractRegistry - Provider 'ManagedItemProvider' has been removed.
Tests run: 2, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 10.955 sec <<< FAILURE! - in org.eclipse.smarthome.automation.module.timer.internal.RuntimeRuleTest
assert that timerTrigger works(org.eclipse.smarthome.automation.module.timer.internal.RuntimeRuleTest)  Time elapsed: 10.623 sec  <<< FAILURE!
java.lang.AssertionError: 
Expected: is <DISABLED>
     but: was <IDLE>
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
	at org.junit.Assert.assertThat(Assert.java:865)
	at org.junit.Assert.assertThat(Assert.java:832)
	at org.junit.Assert$assertThat.callStatic(Unknown Source)
	at org.eclipse.smarthome.automation.module.timer.internal.RuntimeRuleTest$_assert_that_timerTrigger_works_closure9.doCall(RuntimeRuleTest.groovy:145)
	at org.eclipse.smarthome.automation.module.timer.internal.RuntimeRuleTest$_assert_that_timerTrigger_works_closure9.doCall(RuntimeRuleTest.groovy)
	at sun.reflect.GeneratedMethodAccessor9.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:90)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:324)
	at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:278)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1016)
	at org.codehaus.groovy.runtime.callsite.PogoMetaClassSite.call(PogoMetaClassSite.java:39)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:45)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:108)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:112)
	at org.eclipse.smarthome.test.OSGiTest.waitForAssert(OSGiTest.groovy:181)
	at org.eclipse.smarthome.test.OSGiTest.waitForAssert(OSGiTest.groovy)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite$PogoCachedMethodSiteNoUnwrapNoCoerce.invoke(PogoMetaMethodSite.java:207)
	at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite.callCurrent(PogoMetaMethodSite.java:56)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallCurrent(CallSiteArray.java:49)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:133)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:141)
	at org.eclipse.smarthome.automation.module.timer.internal.RuntimeRuleTest.assert that timerTrigger works(RuntimeRuleTest.groovy:143)

check if timer trigger moduleType is registered(org.eclipse.smarthome.automation.module.timer.internal.RuntimeRuleTest)  Time elapsed: 0.022 sec

Results :

Failed tests: 
  RuntimeRuleTest.assert that timerTrigger works:143->OSGiTest.waitForAssert:-1->OSGiTest.waitForAssert:181 
Expected: is <DISABLED>
     but: was <IDLE>

Tests run: 2, Failures: 1, Errors: 0, Skipped: 0
```

**Comment by maggu2810** [comment]
Another point:
It seems that a rule is trigger as much as possible if a timer fulfils the requirements.
Is this expected behaviour? 
If I use a cron timer as trigger and the minimum time unit we can configure are seconds, I would assume that a rule is not triggered multiple times in that second. But it seems that the assumption has been wrong.
If I use e.g.
So it seems that if I use e.g. 6 for hour, 0 for minutes and 0 for seconds (or another cron timer) to trigger e.g. an alarm or something similar the number of executions of the rule is not defined and I must always use a condition to prevent multiple executions?

**Comment by chriskn** [comment]
Is it possible that this is a timing issue related to the performance of the build system?
This test worked for month and still works in our build system and local on my machine.

**Comment by maggu2810** [comment]
@chriskn This test is failing on different systems since a long period from time to time (the Eclipse Hudson build system, Travis CI, ...). See #1757

I am still not sure if a timer that configurable granularity are seconds should trigger multiple times between a second.
Normally cron jobs are not executed more than once in a second why should we do it?
I assume that the "trigger as much as possible" logic eats the whole CPU (which is at least not a good idea for embedded systems).

**Comment by maggu2810** [comment]
Okay, at least the normal quartz usage triggers the job only once a second.

@danchom What causes the "NONE" prints here (I assumed it has been by the executed cron jobs)? Why does the deactivate not work after 10 seconds?

**Comment by maggu2810** [comment]
[log.txt](https://github.com/eclipse/smarthome/files/451412/log.txt)

Some lines of the log:
```text
20:38:04.167 [main] INFO  o.e.s.a.m.t.internal.RuntimeRuleTest - Rule is enabled and idle
20:38:04.169 [main] INFO  o.e.s.a.m.t.internal.RuntimeRuleTest - Disable rule
20:38:04.176 [main] INFO  o.e.s.a.m.t.internal.RuntimeRuleTest - Rule is disabled
20:38:04.178 [main] INFO  o.e.s.a.m.t.internal.RuntimeRuleTest - Try to set lamp item state to OFF
20:38:04.181 [main] INFO  o.e.s.a.m.t.internal.RuntimeRuleTest - Lamp item state is OFF
20:38:04.181 [main] INFO  o.e.s.a.m.t.internal.RuntimeRuleTest - Enable rule and wait for lamp item state is ON
20:38:04.182 [main] INFO  o.e.s.a.core.internal.RuleEngine - The rule rId = MyRule-484698273 is already enabled

10 Seconds later
java.lang.AssertionError: 
Expected: is <ON>
     but: was <OFF>
```

Why is the rule already enabled if the state is disabled before?

@danchom I reported this error on "Jun 28" the first time. Any reason you do never reply to this topic?

**Comment by maggu2810** [comment]
I think the setEnabled method of the rule registry does not work as expected.
I changed the test and the only thing the test does now is to enable and disable the rule very often.
If "setEnabled" is done only once for disable and only once for disable it fails all the time using just enough iterations.
If you uncomment the setEnabled call in the waitForAssert so it tries to do it multiple times, the test succeeds in every time.
So, why can a setEnable call fail?

```groovy
        def numberOfTests = 1000
        for (int i=0; i < numberOfTests;i++) {
            logger.info("Disable rule");
            ruleRegistry.setEnabled(rule.UID, false)
            waitForAssert({
                //ruleRegistry.setEnabled(rule.UID, false)
                final RuleStatusInfo ruleStatus = ruleRegistry.getStatus(rule.UID)
                println "Rule status: " + ruleStatus
                assertThat ruleStatus.status, is(RuleStatus.DISABLED)
            })
            logger.info("Rule is disabled");

            logger.info("Enable rule");
            ruleRegistry.setEnabled(rule.UID, true)
            waitForAssert({
                //ruleRegistry.setEnabled(rule.UID, true)
                final RuleStatusInfo ruleStatus = ruleRegistry.getStatus(rule.UID)
                println "Rule status: " + ruleStatus
                boolean allFine
                if (ruleStatus.status.equals(RuleStatus.IDLE) || ruleStatus.status.equals(RuleStatus.RUNNING)) {
                    allFine = true
                } else {
                    allFine = false
                }
                assertThat allFine, is(true)
            })
            logger.info("Rule is enabled");
        }
```

**Comment by maggu2810** [comment]
I removed the triggers, conditions and actions of the rule.
No failure anymore.

I enabled only the triggers (the one timer trigger), the setEnabled fails again (at least once in 1000 iterations).

@danchom Could the disable or enable fail if the rule is currently running?

**Comment by marinmitev** [comment]
@maggu2810 please note that @danchom is on a sick leave this week.

**Comment by maggu2810** [comment]
@marinmitev Thanks for this information. @danchom Get well soon!

**Comment by danchom** [comment]
Hi maggu2810, Sorry for the late response, but I was seek. I see that there is merged PR for this issue. 
Is still have a problem with setEnable()? I've tested a rule with timer (* * * * * * ?) and it is stopping immediately.

**Comment by maggu2810** [comment]
@danchom I have no problem anymore and the PR could be closed.
I am not sure if I understand your problem. What it is about?

**Comment by maggu2810** [comment]
Fixed by: #2102 and #2126
